### DISCUSSION TITLE
DimWit: `Tensor2[Batch, Feature, Float32]` — semantic tensor types for JAX (Scala 3 frontend, XLA backend, zero overhead)

### DISCUSSION THREAD

**Discussion by benikm91** [jax-ml/jax#40592] [Show and tell]
Because JAX traces by executing code, the code doing the calling doesn't have to be Python. We used that to build DimWit, a Scala 3 tensor library that traces through JAX — and makes axis names static types: `Tensor2[Batch, Feature, Float32]` is a batch of feature vectors, `Tensor3[Height, Width, Channel, UInt8]` is an image, and you write `t.sum(Axis[Batch])` instead of `axis=0`. Every operation names its axes and the compiler verifies them — no `.transpose` to line positions up. Compile errors for wrong axis, missing batch axis, swapped height/width axes.

If you know Haliax or jaxtyping: this is named tensors on steroids. In DimWit the names are static types, meaning axis names are automatically tracked on every tensor and verified through all operations. Of 13 real silent-shape defects we reconstructed from the PyTorch, Transformers, Keras, and vLLM issue trackers, DimWit rejects 12 at compile time.

And it costs nothing at execution: the language bridge is only crossed during tracing, then it's StableHLO + XLA like any JAX program. A 124M GPT-2 written entirely in DimWit trains at native-JAX throughput (~43k vs ~44k tok/s, same node).

`grad`, `jit`, `vmap`, `pytrees`, and explicit PRNG keys all carry over, but with types. Here is a linear regression example (see repository for more advanced examples like GPT-2):

```scala
case class Params(A: Tensor1[In, Float32], b: Tensor0[Float32]) // Params is a tensor tree

case class LinearRegression(params: Params):
  def apply(x: Tensor1[In, Float32]): Tensor0[Float32] =
    params.A.dot(Axis[In])(x) + params.b

def costFnFor(
  X: Tensor2[Sample, In, Float32],
  y: Tensor1[Sample, Float32]
)(params: Params): Tensor0[Float32] =
  val model = LinearRegression(params)
  val yHat = X.vmap(Axis[Sample])(model)
  (y - yHat).pow(2).mean

val dCost: Params => Params = grad(costFnFor(X, y))
val nextParams = params -- alpha * dCost(params) // -- marks minus over tree structure
```

Would love to hear what you think. Any feedback and questions are more than welcome. And a GitHub star means a lot at this early stage.

DimWit (JAX wrapper): https://github.com/dimwit-dev/dimwit
DeepWit (functional deep learning library on DimWit): https://github.com/dimwit-dev/deepwit

**Comment by nh66bp9nvw-sudo** [comment]
Interesting ✨
